File: src/main/java/com/example/parking/domain/reservation/service/ReservationService.java
    package com.example.parking.domain.reservation.service;

    import com.example.parking.domain.parkingLot.entity.ParkingLot;
    import com.example.parking.domain.parkingLot.repository.ParkingLotRepository;
    import com.example.parking.domain.parkingspot.dto.ParkingSpotDto;
    import com.example.parking.domain.parkingspot.entity.ParkingSpot;
    import com.example.parking.domain.parkingspot.entity.SpotStatus;
    import com.example.parking.domain.parkingspot.repository.ParkingSpotRepository;
    import com.example.parking.domain.parkingspot.service.ParkingSpotService;
    import com.example.parking.domain.payment.repository.PaymentRepository;
    import com.example.parking.domain.reservation.dto.ReservationReqDto;
    import com.example.parking.domain.reservation.dto.ReservationResDto;
    import com.example.parking.domain.reservation.entity.Reservation;
    import com.example.parking.domain.reservation.entity.ReservationStatus;
    import com.example.parking.domain.reservation.repository.ReservationRepository;
    import com.example.parking.domain.user.entity.User;
    import com.example.parking.domain.user.repository.UserRepository;
    import com.example.parking.global.sse.SseEmitterManager;
    import lombok.RequiredArgsConstructor;
    import lombok.extern.slf4j.Slf4j;
    import org.springframework.beans.factory.ObjectProvider;
    import org.springframework.scheduling.TaskScheduler;
    import org.springframework.stereotype.Service;
    import org.springframework.transaction.annotation.Transactional;

    import java.time.Instant;
    import java.time.LocalDateTime;
    import java.util.List;
    import java.util.stream.Collectors;

    @Slf4j
    @Service
    @RequiredArgsConstructor
    @Transactional(readOnly = true)
    public class ReservationService {

        private final UserRepository userRepository;
        private final ParkingLotRepository parkingLotRepository;
        private final ParkingSpotRepository parkingSpotRepository;
        private final ParkingSpotService parkingSpotService;
        private final ReservationRepository reservationRepository;
        private final TaskScheduler taskScheduler; //1. TaskScheduler 주입
        private final ObjectProvider<ReservationService> reservationServiceProvider;
        private final PaymentRepository paymentRepository;
        private final SseEmitterManager sseEmitterManager;

        // [CUS-04] 예약 관리 - 내 예약 목록 조회
        public List<ReservationResDto> getMyReservations(Long userId, ReservationStatus status) {
            // Repository의 findAllByUserIdWithDetails 쿼리에 status != 'CANCELLED' 조건이 필요합니다.
            return reservationRepository.findAllByUserIdWithDetails(userId, status).stream()
                    .map(ReservationResDto::from)
                    .collect(Collectors.toList());
        }

        // [CUS-04] 예약 관리 - 내 특정 예약 상세 조회
        public ReservationResDto getReservationDetail(Long reservationId, Long userId) {
            // 이미 취소된 예약은 조회가 되지 않도록 Repository에서 필터링하거나 여기서 상태를 체크합니다.
            Reservation reservation = reservationRepository.findByIdAndUserIdWithDetails(reservationId, userId)
                    .orElseThrow(() -> new IllegalArgumentException("존재하지 않거나 권한이 없는 예약입니다."));

            return ReservationResDto.from(reservation);
        }

        @Transactional
        public void cancelReservation(Long reservationId, Long userId, boolean isForced) {
            //Fetch Join으로 자리 정보까지 한 번에 가져옵니다.
            Reservation reservation = reservationRepository.findByIdWithParkingSpot(reservationId)
                    .orElseThrow(() -> new IllegalArgumentException("존재하지 않는 예약입니다."));

            if (reservation.getStatus() == ReservationStatus.CANCELED) {
                throw new IllegalStateException("이미 취소 처리된 예약입니다.");
            }

            //[검증] 관리자가 아닐 때만(isForced=false) 본인 확인 및 30분 정책 체크
            if (!isForced) {
                // 1. 권한 검증: 본인 예약인지 확인
                if (userId == null || !reservation.getUser().getId().equals(userId)) {
                    throw new IllegalArgumentException("해당 예약을 취소할 권한이 없습니다.");
                }

                // 2. 시간 검증: 입차 30분 전까지만 취소 가능
                LocalDateTime now = LocalDateTime.now();
                if (now.isAfter(reservation.getStartTime().minusMinutes(30))) {
                    throw new IllegalStateException("입차 30분 전까지만 취소가 가능합니다.");
                }
            }

            if (reservation.getStatus() == ReservationStatus.CONFIRMED) {
                paymentRepository.findByReservationId(reservationId).ifPresent(payment -> {
                    payment.refund(); // Payment 엔티티의 상태를 REFUND로 변경
                    log.info("[환불 처리] 예약 ID: {} - 취소 정책에 따른 환불이 완료되었습니다.", reservationId);
                });
            }

            // 4. 상태 변경 및 자리 반환
            reservation.cancel();

            ParkingSpot spot = reservation.getParkingSpot();
            if (spot.getStatus() != SpotStatus.AVAILABLE) {
                spot.release();
                sseEmitterManager.notify(
                        spot.getParkingLot().getId(),
                        new ParkingSpotDto(spot)
                );
            }
        }

        // [CUS-03] 예약 생성
        @Transactional
        public ReservationResDto createReservation(Long userId, ReservationReqDto reqDto) {
            // 1. DTO에서 안전하게 파싱된 시간을 가져옵니다.
            LocalDateTime start = reqDto.getParsedStartTime();
            LocalDateTime end = reqDto.getParsedEndTime();

            // 2. 시간 유효성 검증
            if (start.isBefore(LocalDateTime.now())) {
                throw new IllegalArgumentException("과거 시간으로 예약할 수 없습니다.");
            }
            if (end.isBefore(start)) {
                throw new IllegalArgumentException("종료 시간이 시작 시간보다 앞설 수 없습니다.");
            }

            // 3. 유저와 주차장 정보 조회
            User user = userRepository.findById(userId)
                    .orElseThrow(() -> new IllegalArgumentException("존재하지 않는 유저입니다."));
            ParkingLot parkingLot = parkingLotRepository.findById(reqDto.parkingLotId)
                    .orElseThrow(() -> new IllegalArgumentException("존재하지 않는 주차장입니다."));

            ParkingSpot parkingSpot = parkingSpotRepository.findById(reqDto.parkingSpotId)
                .orElseThrow(() -> new IllegalArgumentException("존재하지 않는 주차 자리입니다."));


            //수정된 부분 1: 현재 자리가 누군가 결제 중(OCCUPIED)인지 먼저 확인
            if (parkingSpot.getStatus() == SpotStatus.OCCUPIED) {
                throw new IllegalStateException("현재 다른 사용자가 결제 진행 중인 자리입니다. 잠시 후 다시 시도해주세요.");
            }

            // 4. 선택한 주차장의 ID와 실제 주차 자리가 속한 주차장의 ID가 일치하는지 검증
            if (!parkingSpot.getParkingLot().getId().equals(reqDto.parkingLotId)) {
                throw new IllegalArgumentException("선택하신 주차장(ID: " + reqDto.parkingLotId +
                        ")에 해당 주차 자리(ID: " + reqDto.parkingSpotId + ")가 존재하지 않습니다.");
            }

            // 4-1. 주차 자리와 내 차종이 서로 다를때의 방어로직.
            if (!parkingSpot.getType().name().equals(user.getVehicleType().name())) {
                throw new IllegalStateException("해당 자리는 " + parkingSpot.getType() +
                        " 전용입니다. 고객님의 차종(" + user.getVehicleType() +
                        ")은 이용할 수 없습니다.");
            }

            // 5. 예약 시간 중복 검사
            long overlapCount = reservationRepository.countOverlappingReservations(
                    parkingSpot.getId(), start, end
            );

            if (overlapCount > 0) {
                throw new IllegalStateException("해당 시간에 이미 예약된 자리입니다.");
            }

            // [신규 추가] 1인 1주차장 1자리 제한 검증
            List<ReservationStatus> activeStatuses = List.of(ReservationStatus.PENDING, ReservationStatus.CONFIRMED, ReservationStatus.COMPLETED);
            boolean hasActiveReservation = reservationRepository.existsByUserIdAndParkingLotIdAndStatusIn(
                userId, reqDto.parkingLotId, activeStatuses
            );

            if (hasActiveReservation) {
                throw new IllegalStateException("이미 이 주차장에 진행 중인 예약(또는 선점)이 존재합니다. 1주차장 당 1자리만 이용 가능합니다.");
            }
            // 6.CAS로 원자적 점유 시도
            int updated = parkingSpotRepository.tryReserve(parkingSpot.getId(), LocalDateTime.now());
            if (updated == 0) {
                throw new IllegalStateException("방금 다른 사용자가 선점했습니다. 다른 자리를 선택해주세요.");
            }

            // 7. CAS 성공 저장용 재조회 (영속 컨텍스트 문제 해결. CAS는 영속 컨텍스트를 비워버리므로)
            ParkingSpot spot = parkingSpotRepository.findById(reqDto.parkingSpotId)
                .orElseThrow(() -> new IllegalArgumentException("존재하지 않는 주차 자리입니다."));

            // CAS 성공 후 SSE 알림 직접 발송
            sseEmitterManager.notify(spot.getParkingLot().getId(), new ParkingSpotDto(spot));


            // 6. 예약 엔티티 생성 및 저장
            Reservation newReservation = Reservation.of(
                    user,
                    parkingLot,
                    spot,
                    start,
                    end
            );

            Reservation savedReservation = reservationRepository.save(newReservation);
            Long reservationId = savedReservation.getId(); // ID 추출

            //2. [실시간 취소 예약] 5분 뒤에 아래 cancelIfUnpaid 메서드를 실행
            taskScheduler.schedule(() -> {
                 //자기 자신의 프록시를 가져와서 호출해야 @Transactional이 정상 작동
                ReservationService self = reservationServiceProvider.getObject();
                self.cancelIfUnpaid(reservationId);
            }, Instant.now().plusSeconds(300));
            log.info("[예약 생성] 50초 타이머 작동 시작 - 예약 ID: {}", reservationId);

            return ReservationResDto.from(savedReservation);
        }

        @Transactional // 반드시 별도의 트랜잭션으로 실행되어야 함
        public void cancelIfUnpaid(Long reservationId) {
            //findById 대신 새로 만든 Fetch Join 메서드 사용
            reservationRepository.findByIdWithParkingSpot(reservationId).ifPresent(res -> {
                if (res.getStatus() == ReservationStatus.PENDING && res.getPaymentRequestedAt() == null) {
                    res.cancel();
                    if (res.getParkingSpot().getStatus() == SpotStatus.OCCUPIED) {
                        res.getParkingSpot().release();
                    }
                    log.info("[1차 선점 취소] 예약 ID: {} - 결제 미진입으로 인한 만료", reservationId);
                } else {
                    // 결제창에 진입한 경우(paymentRequestedAt != null), 2차 타이머(결제팀)에게 처리를 맡깁니다.
                    log.info("[1차 타이머 종료] 결제 프로세스 확인됨. 예약 ID: {}", reservationId);
                }
            });
        }
        // [Step 2] 결제 프로세스 진입 (결제 팀이 호출)
        @Transactional
        public void startPaymentProcess(Long reservationId) {
            Reservation res = reservationRepository.findByIdWithParkingSpot(reservationId)
                    .orElseThrow(() -> new IllegalArgumentException("존재하지 않는 예약입니다."));

            res.startPayment(); // paymentRequestedAt 기록
            res.getParkingSpot().updateStatus(SpotStatus.PAYING); // OCCUPIED -> PAYING
            log.info("[결제 시작] 예약 ID: {}, 자리 상태: PAYING", reservationId);
        }

        // [Step 3] 결제 최종 성공 (결제 팀이 호출)
        @Transactional
        public void completePayment(Long reservationId) {
            Reservation res = reservationRepository.findByIdWithParkingSpot(reservationId)
                    .orElseThrow(() -> new IllegalArgumentException("존재하지 않는 예약입니다."));

            res.confirm(); // PENDING -> CONFIRMED
            // 실제 상태를 변경하기 전에 체크
            if (res.getParkingSpot().getStatus() == SpotStatus.PAYING) {
                res.getParkingSpot().release();
                log.info("[성공] 주차자리 해제 완료: spotId {}", res.getParkingSpot().getId());
            } else {
                //이미 다른 곳에서 바꿨다면 여기서 로그를 남깁니다.
                log.info("[스킵] 주차자리가 이미 AVAILABLE 상태입니다: spotId {}", res.getParkingSpot().getId());
            }
        }


        private void validateReservationOpenTime() {
            LocalDateTime now = LocalDateTime.now();

            int hour = now.getHour();

            if (hour < 14 || hour >= 24) {
                throw new IllegalStateException("예약은 매일 22시부터 24시까지만 가능합니다.");
            }
        }

    }

File: src/main/java/com/example/parking/domain/reservation/service/ReservationService.kt
package com.example.parking.domain.reservation.service

import com.example.parking.domain.parkingLot.entity.ParkingLot
import com.example.parking.domain.parkingLot.repository.ParkingLotRepository
import com.example.parking.domain.parkingspot.dto.ParkingSpotDto
import com.example.parking.domain.parkingspot.entity.ParkingSpot
import com.example.parking.domain.parkingspot.entity.SpotStatus
import com.example.parking.domain.parkingspot.repository.ParkingSpotRepository
import com.example.parking.domain.parkingspot.service.ParkingSpotService
import com.example.parking.domain.payment.repository.PaymentRepository
import com.example.parking.domain.reservation.dto.ReservationReqDto
import com.example.parking.domain.reservation.dto.ReservationResDto
import com.example.parking.domain.reservation.entity.Reservation
import com.example.parking.domain.reservation.entity.ReservationStatus
import com.example.parking.domain.reservation.repository.ReservationRepository
import com.example.parking.domain.user.entity.User
import com.example.parking.domain.user.repository.UserRepository
import com.example.parking.global.sse.SseEmitterManager
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.ObjectProvider
import org.springframework.scheduling.TaskScheduler
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.transaction.support.TransactionSynchronization
import org.springframework.transaction.support.TransactionSynchronizationManager
import java.time.Instant
import java.time.LocalDateTime
import java.time.Clock

@Service
@Transactional(readOnly = true)
class ReservationService(
    private val userRepository: UserRepository,
    private val parkingLotRepository: ParkingLotRepository,
    private val parkingSpotRepository: ParkingSpotRepository,
    private val parkingSpotService: ParkingSpotService,
    private val reservationRepository: ReservationRepository,
    private val taskScheduler: TaskScheduler,
    private val reservationServiceProvider: ObjectProvider<ReservationService>,
    private val paymentRepository: PaymentRepository,
    private val sseEmitterManager: SseEmitterManager,
    private val clock: Clock = Clock.systemDefaultZone()
) {
    private val log = LoggerFactory.getLogger(ReservationService::class.java)

    // [CUS-04] 내 예약 목록 조회
    fun getMyReservations(userId: Long, status: ReservationStatus?): List<ReservationResDto> =
        reservationRepository.findQAllByUserIdWithDetails(userId, status)
            .map { ReservationResDto.from(it) }

    // [CUS-04] 내 특정 예약 상세 조회
    fun getReservationDetail(reservationId: Long, userId: Long): ReservationResDto {
        val reservation = reservationRepository.findQByIdAndUserId(reservationId, userId)
            ?: throw IllegalArgumentException("존재하지 않거나 권한이 없는 예약입니다.")
        return ReservationResDto.from(reservation)
    }

    // [CUS-04] 예약 취소
    @Transactional
    fun cancelReservation(reservationId: Long, userId: Long?, isForced: Boolean) {
        val reservation = reservationRepository.findQByIdWithParkingSpot(reservationId)
            ?: throw IllegalArgumentException("존재하지 않는 예약입니다.")

        if (reservation.status == ReservationStatus.CANCELED) {
            throw IllegalStateException("이미 취소 처리된 예약입니다.")
        }

        if (!isForced) {
            if (userId == null || reservation.user.id != userId) {
                throw IllegalArgumentException("해당 예약을 취소할 권한이 없습니다.")
            }
            if (LocalDateTime.now(clock).isAfter(reservation.startTime.minusMinutes(30))) {
                throw IllegalStateException("입차 30분 전까지만 취소가 가능합니다.")
            }
        }

        if (reservation.status == ReservationStatus.CONFIRMED) {
            paymentRepository.findByReservationId(reservationId).ifPresent { payment ->
                payment.refund()
                log.info("[환불 처리] 예약 ID: {} - 취소 정책에 따른 환불이 완료되었습니다.", reservationId)
            }
        }

        reservation.cancel()

        val spot = reservation.parkingSpot
        if (spot.status != SpotStatus.AVAILABLE && reservation.startTime.isBefore(LocalDateTime.now(clock))) {
            spot.release()
            val lotId = checkNotNull(spot.parkingLot.id)
            TransactionSynchronizationManager.registerSynchronization(object : TransactionSynchronization {
                override fun afterCommit() {
                    sseEmitterManager.notify(lotId, ParkingSpotDto(spot))
                }
            })
        }
    }

    // [CUS-03] 예약 생성
    @Transactional
    fun createReservation(userId: Long, reqDto: ReservationReqDto): ReservationResDto {
        val start = reqDto.getParsedStartTime()
        val end = reqDto.getParsedEndTime()

        validateTime(start, end)
        validateReservationOpenTime()

        val user = findUser(userId)
        val parkingLot = findParkingLot(reqDto.parkingLotId)
        val parkingSpot = findAndValidateSpot(reqDto, user)

        val spot = reserveSpot(parkingSpot, reqDto.parkingSpotId)

        validateNoOverlap(spot, start, end)
        validateNoActiveReservation(userId, reqDto.parkingLotId)

        val savedReservation = reservationRepository.save(
            Reservation.of(user, parkingLot, spot, start, end)
        )
        val reservationId = checkNotNull(savedReservation.id)
        val lotId = checkNotNull(spot.parkingLot.id)

        registerAfterCommit(reservationId, lotId, spot)

        return ReservationResDto.from(savedReservation)
    }

// ==================== private 메서드 ====================

    private fun validateTime(start: LocalDateTime, end: LocalDateTime) {
        if (start.isBefore(LocalDateTime.now(clock))) {
            throw IllegalArgumentException("과거 시간으로 예약할 수 없습니다.")
        }
        if (end.isBefore(start)) {
            throw IllegalArgumentException("종료 시간이 시작 시간보다 앞설 수 없습니다.")
        }

        val today = LocalDateTime.now(clock).toLocalDate()
        val startDate = start.toLocalDate()
        val daysFromToday = java.time.temporal.ChronoUnit.DAYS.between(today, startDate)

        // 내일(+1일), +4일~+6일만 허용 / +2일, +3일은 차단
        if (daysFromToday == 2L || daysFromToday == 3L) {
            throw IllegalStateException("해당 날짜는 아직 예약이 오픈되지 않았습니다.")
        }

        // +7일 이상은 차단
        if (daysFromToday > 6L) {
            throw IllegalArgumentException("예약은 최대 6일 후까지만 가능합니다.")
        }
    }

    private fun validateReservationOpenTime() {
//        val hour = LocalDateTime.now(clock).hour
//        if (hour < 22) {
//            throw IllegalStateException("예약은 매일 22시부터 24시까지만 가능합니다.")
//        }
    }

    private fun findUser(userId: Long): User =
        userRepository.findById(userId)
            .orElseThrow { IllegalArgumentException("존재하지 않는 유저입니다.") }

    private fun findParkingLot(parkingLotId: Long): ParkingLot =
        parkingLotRepository.findById(parkingLotId)
            .orElseThrow { IllegalArgumentException("존재하지 않는 주차장입니다.") }

    private fun findAndValidateSpot(reqDto: ReservationReqDto, user: User): ParkingSpot {
        val parkingSpot = parkingSpotRepository.findById(reqDto.parkingSpotId)
            .orElseThrow { IllegalArgumentException("존재하지 않는 주차 자리입니다.") }

        if (parkingSpot.status == SpotStatus.OCCUPIED) {
            throw IllegalStateException("현재 다른 사용자가 결제 진행 중인 자리입니다. 잠시 후 다시 시도해주세요.")
        }

        if (parkingSpot.parkingLot.id != reqDto.parkingLotId) {
            throw IllegalArgumentException(
                "선택하신 주차장(ID: ${reqDto.parkingLotId})에 해당 주차 자리(ID: ${reqDto.parkingSpotId})가 존재하지 않습니다."
            )
        }

        val vehicleType = user.vehicleType
            ?: throw IllegalStateException("차량 정보가 등록되지 않았습니다. 차종을 먼저 등록해주세요.")
        if (parkingSpot.type.name != vehicleType.name) {
            throw IllegalStateException(
                "해당 자리는 ${parkingSpot.type} 전용입니다. 고객님의 차종(${vehicleType})은 이용할 수 없습니다."
            )
        }

        return parkingSpot
    }

    private fun reserveSpot(parkingSpot: ParkingSpot, spotId: Long): ParkingSpot {
        val updated = parkingSpotRepository.tryReserve(
            checkNotNull(parkingSpot.id), LocalDateTime.now(clock)
        )
        if (updated == 0) {
            throw IllegalStateException("방금 다른 사용자가 선점했습니다. 다른 자리를 선택해주세요.")
        }
        return parkingSpotRepository.findById(spotId)
            .orElseThrow { IllegalArgumentException("존재하지 않는 주차 자리입니다.") }
    }

    private fun validateNoOverlap(spot: ParkingSpot, start: LocalDateTime, end: LocalDateTime) {
        val overlapCount = reservationRepository.countQOverlappingReservations(
            checkNotNull(spot.id), start, end
        )
        if (overlapCount > 0) {
            throw IllegalStateException("해당 시간에 이미 예약된 자리입니다.")
        }
    }

    private fun validateNoActiveReservation(userId: Long, parkingLotId: Long) {
        val activeStatuses = listOf(
            ReservationStatus.PENDING,
            ReservationStatus.CONFIRMED,
            ReservationStatus.COMPLETED
        )
        if (reservationRepository.existsQByUserIdAndParkingLotIdAndStatusIn(userId, parkingLotId, activeStatuses)) {
            throw IllegalStateException("이미 이 주차장에 진행 중인 예약이 존재합니다. 1주차장 당 1자리만 이용 가능합니다.")
        }
    }

    private fun registerAfterCommit(reservationId: Long, lotId: Long, spot: ParkingSpot) {
        TransactionSynchronizationManager.registerSynchronization(object : TransactionSynchronization {
            override fun afterCommit() {
                sseEmitterManager.notify(lotId, ParkingSpotDto(spot))
                taskScheduler.schedule({
                    val self = reservationServiceProvider.getObject()
                    self.cancelIfUnpaid(reservationId)
                }, Instant.now().plusSeconds(300))
                log.info("[예약 생성] 5분 타이머 작동 시작 - 예약 ID: {}", reservationId)
            }
        })
    }

    @Transactional
    fun cancelIfUnpaid(reservationId: Long) {
        reservationRepository.findQByIdWithParkingSpot(reservationId)?.let { res ->
            if (res.status == ReservationStatus.PENDING && res.paymentRequestedAt == null) {
                res.cancel()
                if (res.parkingSpot.status == SpotStatus.OCCUPIED) {
                    res.parkingSpot.release()
                }
                log.info("[1차 선점 취소] 예약 ID: {} - 결제 미진입으로 인한 만료", reservationId)
            } else {
                log.info("[1차 타이머 종료] 결제 프로세스 확인됨. 예약 ID: {}", reservationId)
            }
        }
    }

    @Transactional
    fun startPaymentProcess(reservationId: Long) {
        val res = reservationRepository.findQByIdWithParkingSpot(reservationId)
            ?: throw IllegalArgumentException("존재하지 않는 예약입니다.")
        res.startPayment()
        res.parkingSpot.updateStatus(SpotStatus.PAYING)
        log.info("[결제 시작] 예약 ID: {}, 자리 상태: PAYING", reservationId)
    }

    @Transactional
    fun completePayment(reservationId: Long) {
        val res = reservationRepository.findQByIdWithParkingSpot(reservationId)
            ?: throw IllegalArgumentException("존재하지 않는 예약입니다.")
        res.confirm()
        if (res.parkingSpot.status == SpotStatus.PAYING) {
            res.parkingSpot.release()
            log.info("[성공] 주차자리 해제 완료: spotId {}", res.parkingSpot.id)
        } else {
            log.info("[스킵] 주차자리가 이미 AVAILABLE 상태입니다: spotId {}", res.parkingSpot.id)
        }
    }
}